R/validate.R

```diff
@@ -188,6 +188,41 @@ new_forecast <- function(data, classname) {
 }
 
 
+#' @title Test Whether An Object Is Of Class `forecast_*`
+#'
+#' @description
+#' Convenience function to test whether an object is of class `forecast_*`.
+#' Specifically, the function tests whether the object matches at least one of
+#' the classes in the `class` argument. By default, it tests whether the object
+#' is of any of the following classes:
+#' - `forecast_binary`
+#' - `forecast_point`
+#' - `forecast_sample`
+#' - `forecast_quantile`
+#'
+#' @inheritParams get_forecast_counts
+#' @param class A character vector indicating the classes to be tested for.
+#' The test passes if the object has at least one of the classes specificed
+#' in the `class` vector.
+#' @return `TRUE` if the object matches one of the specified classes,
+#' `FALSE` otherwise.
+#' @export
+#' @keywords check-forecasts
+#' @examples
+#' forecast_binary <- as_forecast(example_binary)
+#' is_forecast(forecast_binary)
+#' is_forecast(forecast_binary, class = "forecast_binary")
+#' is_forecast(forecast_binary, class = c("forecast_binary", "forecast_point"))
+is_forecast <- function(data,
```

Comment:
why isn't this just an s3 method?

Reply:
oh hm. Good question. Would you have this be just a `is_forecast.default()` method? Or a separate `is_forecast.forecast_binary()`, `is_forecast.forecast_point()` etc.? 
Realistically the code for them would all be the same, wouldn't it? 

I was also wondering whether there is a case where you'd want to run a check like "is this either a binary or a point forecast"? With the current implementation, you could do that. If we just have an `is_forecast()` S3 method, then that would also pass for a quantile forecast, for example.

Reply:
What do you think about the use case to test whether something is of class a or b, but not c or d? Or testing if something is e.g. of class `forecast_binary` specifically?

Reply:
So as implemented it can't be extended outside of `scoringutils` so breaks the philosophy of the redesign.

If you instead implement it as: 

`is_forecast.default(return(FALSE))`, `is_forecast.forecast_binary(return(TRUE))` Then it can and the lower level methods can also be used to check for specific (and multiple types). 

I agree its a bit infra heavy but then it also saves having some magic strings.

Reply:
I think I'd be fine with this. 
Checking for multiple classes would then be something like 
`is_forecast.forecast_binary(object) | is_forecast.forecast_point(object)`?

Reply:
I just realised that 
```
is_forecast.forecast_binary(object) | is_forecast.forecast_point(object)
```
doesn't work - the function only ever returns `TRUE` if I implement it as 

```
is_forecast <- function(object) {
  UseMethod("is_forecast")
}

is_forecast.default <- function(object, ...) {
  return(FALSE)
}

is_forecast.forecast_binary <- function(object, ...) {
  return(TRUE)
}
```

Reply:
There is the ugly version: 

```
is_forecast <- function(data,
                        class = c("forecast_binary", "forecast_point",
                                  "forecast_sample", "forecast_quantile"), 
                        allowed = c("forecast_binary", "forecast_point",
                                    "forecast_sample", "forecast_quantile")) {
  allowed <- c("forecast_binary", "forecast_point",
               "forecast_sample", "forecast_quantile")
  classes <- match.arg(class, allowed, several.ok = TRUE)
  return(inherits(data, classes))
}
```
where `allowed` would be there to do `match.arg()`

and the lean version which would just be 

```
is_forecast <- function(data,
                        class = c("forecast_binary", "forecast_point",
                                  "forecast_sample", "forecast_quantile")) {
  return(inherits(data, classes))
}
```

Reply:
Can't you check it inherits the right class in the return in the same was you do in the above? For the auto-dispatch this will be true but will be meaningful when used directly

Reply:
One would imagine I should have been able to arrive at that solution myself 🤷 :-D